src/mqueryfront/src/RecentPage.js

```diff
@@ -1,17 +1,182 @@
+        this.handleClose = this.handleClose.bind(this);
+        this.handleFilter = this.handleFilter.bind(this);
+        this.getHead = this.getHead.bind(this);
+        this.getDistinctList = this.getDistinctList.bind(this);
+        this.handlePageChange = this.handlePageChange.bind(this);
+        this.getJobsPage = this.getJobsPage.bind(this);
+    }
+
+    handlePageChange(pageNumber) {
+        this.setState({ activePage: pageNumber });
+    }
+
+    handleCancelJob(id) {
+        const { jobs } = this.state;
+        const index = jobs.findIndex((obj) => obj.id === id);
+
+        if (index >= 0) {
+            const newJobs = [...jobs];
+            newJobs[index].status = "cancelled";
+
+            axios.delete(API_URL + "/job/" + id).then((response) => {
+                this.setState({ jobs: newJobs, head: this.getHead(newJobs) });
+            });
+        }
+    }
+
+    handleClose(id) {
+        const { jobs } = this.state;
+        const index = jobs.findIndex((obj) => obj.id === id);
+
+        if (index >= 0) {
+            const newJobs = [...jobs.slice(0, index), ...jobs.slice(index + 1)];
+
+            this.setState({ jobs: newJobs, head: this.getHead(newJobs) });
+        }
+    }
+
+    handleFilter(name, value, index) {
+        const { filter } = this.state;
+
+        if (index > 0) {
+            this.setState({
+                filter: { name: name, value: value },
+                activePage: 1,
+            });
+        } else {
+            if (filter && filter.name === name)
+                this.setState({ filter: null, activePage: 1 });
+        }
+    }
+
+    getHead(jobs) {
+        const head = [
+            {
+                title: "Job name",
+                attrubuteName: "rule_name",
+                valueList: this.getDistinctList(jobs, "rule_name"),
+            },
+            {
+                title: "Author",
+                attrubuteName: "rule_author",
+                valueList: this.getDistinctList(jobs, "rule_author"),
+            },
+            {
+                title: "Matches",
+                attrubuteName: "files_matched",
+                valueList: this.getDistinctList(jobs, "files_matched", true),
+            },
+            {
+                title: "Status/Progress",
+                attrubuteName: "status",
+                valueList: this.getDistinctList(jobs, "status"),
+            },
+            {
+                title: "Actions",
+                attrubuteName: "",
+                valueList: null,
+            },
+        ];
+
+        return head;
+    }
+
+    getDistinctList(
+        arrayOfObjects,
+        attributeName,
+        sortNumerically = false,
+        allElement = "All"
+    ) {
+        let returnList = null;
+
+        if (arrayOfObjects) {
+            returnList = arrayOfObjects
+                .map((item) => item[attributeName])
+                .filter((value, index, self) => self.indexOf(value) === index)
+                .sort();
```

Comment:
This sort is not needed now, right?

```suggestion
                .filter((value, index, self) => self.indexOf(value) === index);
```

Reply:
Of course!